src/workerd/api/sockets.c++

```diff
@@ -128,4 +134,38 @@ void Socket::handleProxyStatus(
   result.markAsHandled();
 }
 
+void Socket::handleReadableEof(jsg::Lock& js) {
+  // Listen for EOF on the ReadableStream.
+  KJ_ASSERT_NONNULL(readable->eofResolverPair).promise.then(js,
+      JSG_VISITABLE_LAMBDA((ref=JSG_THIS), (ref), (jsg::Lock& js) {
+    return ref->maybeCloseWriteSide(js);
+  })).markAsHandled();
+}
+
+jsg::Promise<void> Socket::maybeCloseWriteSide(jsg::Lock& js) {
+  // When `allowHalfOpen` is set to true then we do not automatically close the write side on EOF.
+  // The default value for `allowHalfOpen` is also false.
+  KJ_IF_MAYBE(opts, options) {
+    if (opts->allowHalfOpen) {
+      return js.resolvedPromise();
+    }
+  }
+
+  // We want to close the socket, but only after its WritableStream has been flushed. We do this
+  // below by calling `close` on the WritableStream which ensures that any data pending on it
+  // is flushed. Then once the `close` either completes or fails we can be sure that any data has
+  // been flushed.
+  return writable->getController().close(js).catch_(js,
+      [](jsg::Lock& js, workerd::jsg::V8Ref<v8::Value> exc) {
+    // A failure to close the WritableStream can indicate one of these things:
```

Comment:
Isn't there a third reason: An actual I/O error occurred? In that case, do we really want to silently ignore the error?